src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -14724,3 +14717,138 @@ void Unit::Whisper(uint32 textId, Player* target, bool isBossWhisper /*= false*/
+
+    //! Cannot strafe left and right at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_STRAFE_LEFT) && mi->HasMovementFlag(MOVEMENTFLAG_STRAFE_RIGHT));
+
+    //! Cannot pitch up and down at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_PITCH_UP) && mi->HasMovementFlag(MOVEMENTFLAG_PITCH_DOWN));
+
+    //! Cannot move forwards and backwards at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_FORWARD) && mi->HasMovementFlag(MOVEMENTFLAG_BACKWARD));
+
+    //! Cannot fly ('flying' flag) without the 'canFly' flag active
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_FLYING) && !mi->HasMovementFlag(MOVEMENTFLAG_CAN_FLY));
+
+    ////! Check for synchronisation between client and server with those server-controlled flags:
+    // CanFly, Hover, Waterwalking, Root (or if necessary PendingRoot), DisableGravity, CanTransitionBetweenSwimAndFly and FeatherFall
+    MovementInfo oldMovementInfo = GetMovementInfo();
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_CAN_FLY) != oldMovementInfo.HasMovementFlag(MOVEMENTFLAG_CAN_FLY));
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_HOVER) != oldMovementInfo.HasMovementFlag(MOVEMENTFLAG_HOVER));
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_WATERWALKING) != oldMovementInfo.HasMovementFlag(MOVEMENTFLAG_WATERWALKING));
+    CHECK_FOR_VIOLATING_FLAGS((mi->HasMovementFlag(MOVEMENTFLAG_ROOT) || mi->HasMovementFlag(MOVEMENTFLAG_PENDING_ROOT)) != (oldMovementInfo.HasMovementFlag(MOVEMENTFLAG_ROOT) || oldMovementInfo.HasMovementFlag(MOVEMENTFLAG_PENDING_ROOT)));
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_DISABLE_GRAVITY) != oldMovementInfo.HasMovementFlag(MOVEMENTFLAG_DISABLE_GRAVITY));
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasExtraMovementFlag(MOVEMENTFLAG2_CAN_TRANSITION_BETWEEN_SWIM_AND_FLY) != oldMovementInfo.HasExtraMovementFlag(MOVEMENTFLAG2_CAN_TRANSITION_BETWEEN_SWIM_AND_FLY));
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_FALLING_SLOW) != oldMovementInfo.HasMovementFlag(MOVEMENTFLAG_FALLING_SLOW));
+
+    // If a root is pending, the unit should be falling
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_PENDING_ROOT) && !mi->HasMovementFlag(MOVEMENTFLAG_FALLING));
+
+    // @todo: if the unit is falling, it should be away from the ground
+
+    // @todo: if a unit is away from the ground and doesn't have Flying or DisableGravity, it should be falling (unless hover)
+
+    //! Cannot fly and fall at the same time
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_FLYING | MOVEMENTFLAG_DISABLE_GRAVITY) && mi->HasMovementFlag(MOVEMENTFLAG_FALLING));
+
+    CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_SPLINE_ENABLED) && (!movespline->Initialized() || movespline->Finalized()));
```

Comment:
Cause of disconnects if using charge in water or seating on vehicle while mounted.

Reply:
mistake , should be : 
```
CHECK_FOR_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_SPLINE_ENABLED) &&
        (!movespline->Initialized() || movespline->Finalized()), MOVEMENTFLAG_SPLINE_ENABLED);
```

Fixed Fear first step. Unit staying one 1 place, and running by path very very faster. =)
Also working charge in water, and charge from land in water.

Reply:
I didn't add this check. I only made it mandatory to succeed. This means that the check itself is wrong. 

I have a limited knowledge about splines though. A second opinion would be nice on this.